storage/SMAIndexSubBlock.cpp

```diff
@@ -0,0 +1,670 @@
+      DCHECK(equal_comparisons_[attr_typeid] == nullptr);
+      less_comparisons_[attr_typeid]
+          = ComparisonFactory::GetComparison(ComparisonID::kLess)
+              .makeUncheckedComparatorForTypes(attribute_type, attribute_type);
+      equal_comparisons_[attr_typeid]
+          = ComparisonFactory::GetComparison(ComparisonID::kEqual)
+              .makeUncheckedComparatorForTypes(attribute_type, attribute_type);
+    }
+
+    // Map the attribute's id to its entry's index.
+    attribute_to_entry_[attribute] = indexed_attribute_num;
+
+    if (!header_->index_consistent) {
+      resetEntry(entry, attribute, attribute_type);
+    } else {
+      // If the data held by min/max is out of line, we must retrieve it as we initialize.
+      if (!TypedValue::RepresentedInline(entry->type_id)) {
+        // First, set to 0 so that we don't try to free invalid memory on copy.
+        // Next, copy from the tuple store. This will copy and give us ownership of out of line data.
+        if (entry->min_entry_ref.valid) {
+          new (&entry->min_entry_ref.value) TypedValue();
+          entry->min_entry_ref.value = tuple_store
+              .getAttributeValueTyped(entry->min_entry_ref.entry_ref_tuple, entry->attribute);
+          entry->min_entry_ref.value.ensureNotReference();
+        }
+        if (entry->max_entry_ref.valid) {
+          new (&entry->max_entry_ref.value) TypedValue();
+          entry->max_entry_ref.value = tuple_store
+              .getAttributeValueTyped(entry->max_entry_ref.entry_ref_tuple, entry->attribute);
+          entry->max_entry_ref.value.ensureNotReference();
+        }
+      }
+    }
+  }
+  initialized_ = true;
+}
+
+SMAIndexSubBlock::~SMAIndexSubBlock() {
+  // Any typed values which have out of line data must be cleared.
+  freeOutOfLineData();
+
+  // Clear any operators which have been allocated.
+  for (int i = 0; i < kNumTypeIDs; ++i) {
+    if (add_operations_[i] != nullptr) {
+      delete add_operations_[i];
+    }
+    if (equal_comparisons_[i] != nullptr) {
+      delete equal_comparisons_[i];
+    }
+    if (less_comparisons_[i] != nullptr) {
+      delete less_comparisons_[i];
+    }
+  }
+
+  // Delete the operator arrays.
+  delete[] add_operations_;
+  delete[] equal_comparisons_;
+  delete[] less_comparisons_;
+  delete[] attribute_to_entry_;
+}
+
+void SMAIndexSubBlock::resetEntry(SMAEntry *entry,
+                                  attribute_id attribute,
+                                  const Type &attribute_type) {
+  entry->attribute = attribute;
+  entry->type_id = attribute_type.getTypeID();
+
+  // Clearing min/max will free any out of line data.
+  if (entry->min_entry_ref.valid) {
+    entry->min_entry_ref.value.clear();
+    entry->min_entry_ref.valid = false;
+  }
+  if (entry->max_entry_ref.valid) {
+    entry->max_entry_ref.value.clear();
+    entry->max_entry_ref.valid = false;
+  }
+  sma_internal::initializeTypedValueForSum(entry);
```

Comment:
The initialized_ flag is set to true when the comparators have been created, so it does not need to be reset when the entries are reset. I will add a comment explaining this.